lib/spack/spack/test/stage.py

```diff
@@ -411,23 +412,29 @@ def check_stage_dir_perms(prefix, path):
     # Ensure the path's subdirectories -- to `$user` -- have their parent's
     # perms while those from `$user` on are owned and restricted to the
     # user.
-    status = os.stat(prefix)
-    gid = status.st_gid
-    uid = os.getuid()
+    assert path.startswith(prefix)
+
     user = getpass.getuser()
-    parts = path[len(prefix) + 1:].split(os.path.sep)
-    have_user = False
-    for part in parts:
-        if part == user:
-            have_user = True
-        prefix = os.path.join(prefix, part)
-        prefix_status = os.stat(prefix)
-        if not have_user:
-            assert gid == prefix_status.st_gid
-            assert status.st_mode == prefix_status.st_mode
-        else:
-            assert uid == status.st_uid
-            assert status.st_mode & stat.S_IRWXU == stat.S_IRWXU
+    prefix_status = os.stat(prefix)
+    uid = os.getuid()
+
+    # Skip processing prefix ancestors since no guarantee they will be in the
+    # required group (e.g. $TEMPDIR on HPC machines).
+    skip = prefix if prefix.endswith(os.sep) else os.path.join(prefix, os.sep)
+    group_paths, entry, user_paths = partition(path.replace(skip, ""), user)
+    if entry != '':
```

Comment:
I modeled the path partition function after the Python `partition` function since:
1) it is intended to be a general function; and
2) there _may_ be future use cases where the caller wants the matching path node to be associated with the ancestors or processed differently (than its ancestors or descendants).

Changed the variable name to hopefully add clarify and added a comment.